Fix publishing of coverage results to coveralls.io (+ misc test config updates)

lukpueh opened the pull request:
**Fixes issue #**:
None

**Description of the changes being introduced by the pull request**:
Code coverage results from our Travis builds have not been published to coveralls.io in a while, due to the required `coveralls` publishing tool not being properly installed. 

Besides fixing publishing of coverage results, this PR cleans up/ updates other testing configurations:
 - Update travis to run on xenial
 - Add a build matrix to run each tox env in a corresponding travis env as per travis/tox best practices.
      https://docs.travis-ci.com/user/languages/python/#using-tox-as-the-build-script
 - Enable Python 3.5 tests, after all we claim to support Python 3.5
 - Remove "only build certain branch" restrictions
 - Use "install" instead of "before_script"  to install dependencies.
    Explicitly listing "install" prevents Travis from automatically running `pip install -r requirements.txt`, which is not necessary because most of those requirements are installed again in each  tox environment.
 - Move pylint and bandit calls to tox (pylint requires runtime dependencies to be installed).


**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Instruct travis to install coveralls

Coveralls is used to publish coverage results online via
coveralls.io.

Travis is already configured to run it "after_success", but this
has failed for a while, because it was not installed.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Update travis to xenial and misc updates

- Add a build matrix to run each tox env in a corresponding
  travis env as per travis/tox best practices.
  https://docs.travis-ci.com/user/languages/python/#using-tox-as-the-build-script
- Add Python 3.5 tests
- Remove only build on certain branch restrictions
- Use "install" instead of "before_script"  to install dependencies.
  Explicitly listing "install" prevents Travis from automatically
  running `pip install -r requirements.txt`, which is not necessary
  because most of those requirements are installed again in each
  tox environment.
- Move pylint and bandit calls to tox (pylint requires
  dependencies) to be installed.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
~**WIP! don't merge!**~

~Coveralls does run now, but it has problems reading the `.coverage` report...~

```
...
coverage.misc.CoverageException: Couldn't read data from '/home/travis/build/theupdateframework/tuf/tests/.coverage': CoverageException: Doesn't seem to be a coverage.py data file
...
coverage.misc.CoverageException: Couldn't read data from '/home/travis/build/theupdateframework/tuf/tests/.coverage': UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa8 in position 31: invalid start byte
```

Commit:
Pin coverage dependency for coveralls

We install coverage inside tox builds to generate test coverage
reports. These reports need to be created with a version supported
by coveralls, which we use (outside of tox) to publish coverage
reports to coveralls.io.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Fix coverage file paths by patching sys.path

This replicates behavior of unittest's `discover` method, and
allows `coverage` and the tool that posts coverage reports to
coveralls.io, i.e. `coveralls`, to record the correct paths and
left-strip the parts leading to the project directory.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
Fixed! See most recent [Travis builds](https://travis-ci.org/theupdateframework/tuf/builds/551203687) and corresponding [coverage reports on coveralls](https://coveralls.io/builds/24238801).

The badge on the main README should be updated once this PR is merged and IIUC we will see GitHub status (as in the table at the bottom) in subsequent PRs, because coveralls needs a reference build.

lukpueh commented:
See most recent [Travis builds](https://travis-ci.org/secure-systems-lab/securesystemslib/builds/551256979?utm_source=github_status&utm_medium=notification) and corresponding [coverage reports on coveralls](https://coveralls.io/builds/24241109).

The badge on the main README should be updated once this PR is merged and IIUC we will see GitHub status (as in the table at the bottom) in subsequent PRs, because coveralls needs a reference build.

tanishqjasoria reviewed (approved):
I went through the PR and it looks good to me!

Commit:
Install TUF in editable mode in tox builds

This commit partially reverts the workaround introduced by
157167e0cc8c3f9bd9ea8d96f1d796e08bfa932d. Instead of patching the
path, we configure tox to install TUF in editable mode, which makes
the tests run against the same files as if the path were patched.
This makes it so that coverage records paths that it can then
normalize when sending them to coveralls.io (see .travis.yml).

See https://github.com/uptane/uptane/pull/195/commits/af227011401287c934045b6d9bef347ef0be0845
for detailed background information.

As a consequence we can now skip building of sdist and installing it
into a virtual env directory in tox.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

Commit:
Move `omit` to `run` section in .coveragerc

Coverage used to be configured to omit certain directories while
reporting.

This commit slightly optimizes coverage to already omit those
directories while measuring coverage.

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

adityasaky commented on .travis.yml:
```diff
@@ -1,29 +1,29 @@
+  # FIXME: Consider refactoring the tests to not require the test aggregation
+  # script being invoked from the `tests` directory, so that `.coverage` is
+  # written to and .coveragrc can also reside in the project root directory, as
+  # as the convention.
```
*is

Commit:
Fix typo in .travis.yml

Signed-off-by: Lukas Puehringer <lukas.puehringer@nyu.edu>

lukpueh commented:
Thanks, @adityasaky!
